In [1]:
import pandas as pd

#import statsmodels.formula.api as smf
#import statsmodels.api as sm

from sklearn.linear_model import LogisticRegression

import numpy as np

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.metrics import confusion_matrix, classification_report

from sklearn.metrics import accuracy_score

from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB

from sklearn.model_selection import train_test_split

from sklearn import neighbors

import warnings
warnings.simplefilter("ignore")

# 1. Data Preparation

## 1.1 Read and Clean Data

    Perform following actions on AimoScore and WeakLinks
    Read,
    Print Lenght
    Remove ? and NAN
    Print Lenght
    Print Summary

In [2]:
url = "../Data/Assignment3/AimoScore_WeakLink_big_scores.xls"
AimoScore = pd.read_excel(url)
print(len(AimoScore))
data = AimoScore[~AimoScore.isin(['?'])]
data = data.dropna(axis=0)
AimoScore = pd.DataFrame(data)
print(len(AimoScore))
AimoScore.describe()

2094
2094


AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2094.000000           2094.000000           2094.000000   
mean      0.572740              0.539248              0.504462   
std       0.236510              0.239284              0.282136   
min       0.010000              0.279770              0.092300   
25%       0.407907              0.279770              0.250149   
50%       0.617248              0.500000              0.500000   
75%       0.759378              0.749761              0.750239   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.560258              0.518065              0.513932   
std                0.217828              0.264104              0.270076   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250239              0.250717   
50%                0.500000              0.500000              0.500478   
75%                0.750239              0.750239              0.750239   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.510336              0.519434              0.635923   
std                0.274473              0.262884              0.153280   
min                0.140124              0.194165              0.303247   
25%                0.250717              0.250717              0.521760   
50%                0.500478              0.500956              0.521760   
75%                0.750239              0.750239              0.749761   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  No_19_NASM_Deviation  No_20_NASM_Deviation  \
count           2094.000000  ...           2094.000000           2094.000000   
mean               0.512548  ...              0.725505              0.715943   
std                0.271193  ...              0.095008              0.100437   
min                0.156385  ...              0.670971              0.656624   
25%                0.250717  ...              0.670971              0.656624   
50%                0.500478  ...              0.670971              0.656624   
75%                0.750239  ...              0.751196              0.751196   
max                1.000000  ...              1.000000              1.000000   

       No_21_NASM_Deviation  No_22_NASM_Deviation  No_23_NASM_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.706319              0.652969              0.710774   
std                0.105886              0.141000              0.103196   
min                0.640401              0.525310              0.648972   
25%                0.642276              0.552846              0.648972   
50%                0.642276              0.552846              0.648972   
75%                0.749761              0.749880              0.751196   
max                1.000000              1.000000              1.000000   

       No_24_NASM_Deviation  No_25_NASM_Deviation  No_1_Time_Deviation  \
count           2094.000000           2094.000000          2094.000000   
mean               0.667517              0.548089             0.512021   
std                0.130989              0.230642             0.273056   
min                0.578192              0.308943             0.148733   
25%                0.578192              0.308943             0.250717   
50%                0.578192              0.500000             0.500478   
75%                0.751196              0.750239             0.750239   
max                1.000000              1.000000             1.000000   

In [3]:
url = "../Data/Assignment3/20190108 scores_and_weak_links.xlsx"
WeakLinks = pd.read_excel(url, engine='openpyxl')
print(len(WeakLinks))
data = WeakLinks[~WeakLinks.isin(['?'])]
data = data.dropna(axis=0)
WeakLinks = pd.DataFrame(data)
print(len(WeakLinks))
WeakLinks.describe()

2096
2096


SCORE  ForwardHead  LeftArmFallForward  RightArmFallForward  \
count  2096.000000  2096.000000         2096.000000          2096.000000   
mean      0.572420     0.301579            0.996842             0.880792   
std       0.236856     0.477732            0.481902             0.578039   
min       0.010000     0.000000            0.000000             0.000000   
25%       0.407843     0.000000            0.690000             0.170000   
50%       0.617248     0.000000            1.120000             1.110000   
75%       0.759416     0.480000            1.390000             1.420000   
max       0.993987     1.500000            1.500000             1.500000   

       LeftShoulderElevation  RightShoulderElevation  ExcessiveForwardLean  \
count            2096.000000             2096.000000           2096.000000   
mean                0.271293                0.465000              0.201150   
std                 0.418118                0.475603              0.419338   
min                 0.000000                0.000000              0.000000   
25%                 0.000000                0.030000              0.000000   
50%                 0.030000                0.325000              0.000000   
75%                 0.390000                0.852500              0.080000   
max                 1.500000                1.500000              1.500000   

       LeftAsymmetricalWeightShift  RightAsymmetricalWeightShift  \
count                  2096.000000                   2096.000000   
mean                      0.175219                      0.065315   
std                       0.397579                      0.252260   
min                       0.000000                      0.000000   
25%                       0.000000                      0.000000   
50%                       0.000000                      0.000000   
75%                       0.050000                      0.000000   
max                       1.500000                      1.500000   

       LeftKneeMovesInward  RightKneeMovesInward  LeftKneeMovesOutward  \
count          2096.000000           2096.000000           2096.000000   
mean              0.037657              0.179089              0.170787   
std               0.140765              0.329131              0.401551   
min               0.000000              0.000000              0.000000   
25%               0.000000              0.000000              0.000000   
50%               0.000000              0.000000              0.000000   
75%               0.000000              0.200000              0.020000   
max               1.310000              1.500000              1.500000   

       RightKneeMovesOutward  LeftHeelRises  RightHeelRises  
count            2096.000000    2096.000000     2096.000000  
mean                0.238912       0.059241        0.102505  
std                 0.528886       0.226736        0.290971  
min                 0.000000       0.000000        0.000000  
25%                 0.000000       0.000000        0.000000  
50%                 0.000000       0.000000        0.000000  
75%                 0.000000       0.000000        0.000000  
max                 1.500000       1.500000        1.500000

### <font color='green'>Data was clean. not found any ? and NAN</font>

# 2 Analysis on IDs

## 2.1 Count IDs

In [4]:
print(len(AimoScore['ID'].value_counts()))
print(len(WeakLinks['ID'].value_counts()))

2091
2096


### <font color='red'>AimoScore has 2094 values but 2091 IDs. which mean 3 rows either have duplicate IDs OR empty</font>
### <font color='green'>WeakLinks is OK</font>

## 2.2 Cross verify IDs
    
    Check which ID of AimoScore are in WeakLinks 
    Check which ID of WeakLinks are in AimoScore

In [5]:
WeakLinks['ID'].isin(AimoScore['ID']).value_counts()

True     2091
False       5
Name: ID, dtype: int64

In [6]:
AimoScore['ID'].isin(WeakLinks['ID']).value_counts()

True    2094
Name: ID, dtype: int64

### <font color='red'>AimoScore has 2094 values matched to WeakLinks. which mean either WeakLinks have empty ID or some IDs in AimoScore are duplicate</font>
### <font color='red'>WeakLinks has 2091 values matched to AimoScore. which mean 5 extra IDs</font>

## 2.3 Find differnt IDs
    
    Check differnet IDs using symmetric_difference and concat

In [7]:
set(AimoScore.ID).symmetric_difference(WeakLinks.ID)

{'076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect',
 '18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect',
 '1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect',
 '49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect',
 'ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect'}

In [8]:
combined_dfs = pd.concat([AimoScore['ID'], WeakLinks['ID']])
combined_dfs.drop_duplicates(keep=False)

277     076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect
809     ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect
953     1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect
1149    18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect
1201    49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect
Name: ID, dtype: object

### <font color='green'>We found 5 IDs that differ</font>

## 2.4 Check Existense of IDs
    Using indexes, and merge, check IDs exist in One OR in both DataFrames

In [9]:
print(AimoScore[AimoScore['ID'] == "076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect"].index)

Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([277], dtype='int64')
Int64Index([809], dtype='int64')
Int64Index([953], dtype='int64')
Int64Index([1149], dtype='int64')
Int64Index([1201], dtype='int64')


In [10]:
AimoScore.merge(WeakLinks, how = 'outer' ,indicator=True).loc[lambda x : x['_merge']=='left_only']

Empty DataFrame
Columns: [AimoScore, No_1_Angle_Deviation, No_2_Angle_Deviation, No_3_Angle_Deviation, No_4_Angle_Deviation, No_5_Angle_Deviation, No_6_Angle_Deviation, No_7_Angle_Deviation, No_8_Angle_Deviation, No_9_Angle_Deviation, No_10_Angle_Deviation, No_11_Angle_Deviation, No_12_Angle_Deviation, No_13_Angle_Deviation, No_1_NASM_Deviation, No_2_NASM_Deviation, No_3_NASM_Deviation, No_4_NASM_Deviation, No_5_NASM_Deviation, No_6_NASM_Deviation, No_7_NASM_Deviation, No_8_NASM_Deviation, No_9_NASM_Deviation, No_10_NASM_Deviation, No_11_NASM_Deviation, No_12_NASM_Deviation, No_13_NASM_Deviation, No_14_NASM_Deviation, No_15_NASM_Deviation, No_16_NASM_Deviation, No_17_NASM_Deviation, No_18_NASM_Deviation, No_19_NASM_Deviation, No_20_NASM_Deviation, No_21_NASM_Deviation, No_22_NASM_Deviation, No_23_NASM_Deviation, No_24_NASM_Deviation, No_25_NASM_Deviation, No_1_Time_Deviation, No_2_Time_Deviation, EstimatedScore, ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises, _merge]
Index: []

[0 rows x 60 columns]

In [11]:
AimoScore.merge(WeakLinks, how = 'outer' ,indicator=True).loc[lambda x : x['_merge']=='right_only']

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
2094        NaN                   NaN                   NaN   
2095        NaN                   NaN                   NaN   
2096        NaN                   NaN                   NaN   
2097        NaN                   NaN                   NaN   
2098        NaN                   NaN                   NaN   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
2094                   NaN                   NaN                   NaN   
2095                   NaN                   NaN                   NaN   
2096                   NaN                   NaN                   NaN   
2097                   NaN                   NaN                   NaN   
2098                   NaN                   NaN                   NaN   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
2094                   NaN                   NaN                   NaN   
2095                   NaN                   NaN                   NaN   
2096                   NaN                   NaN                   NaN   
2097                   NaN                   NaN                   NaN   
2098                   NaN                   NaN                   NaN   

      No_9_Angle_Deviation  ...  ExcessiveForwardLean  \
2094                   NaN  ...                  0.00   
2095                   NaN  ...                  0.00   
2096                   NaN  ...                  0.00   
2097                   NaN  ...                  0.00   
2098                   NaN  ...                  1.39   

      LeftAsymmetricalWeightShift  RightAsymmetricalWeightShift  \
2094                          0.0                           0.0   
2095                          0.0                           0.0   
2096                          0.0                           0.0   
2097                          0.0                           0.0   
2098                          0.0                           0.0   

      LeftKneeMovesInward  RightKneeMovesInward  LeftKneeMovesOutward  \
2094                  0.0                  0.00                  0.00   
2095                  0.0                  0.00                  0.02   
2096                  0.0                  0.00                  0.02   
2097                  0.0                  0.00                  0.00   
2098                  0.0                  0.01                  0.00   

      RightKneeMovesOutward  LeftHeelRises  RightHeelRises      _merge  
2094                    0.0           0.00            0.00  right_only  
2095                    1.5           0.35            1.04  right_only  
2096                    1.5           0.68            1.14  right_only  
2097                    0.0           0.00            0.00  right_only  
2098                    0.0           0.00            0.00  right_only  

[5 rows x 60 columns]

### <font color='red'>IDs not found in AimoScore</font>
### <font color='green'>WeakLinks has all 5 extra IDs</font>
### <font color='red'>We can conclude that AimoScore has 3 duplicate IDs</font>

## 2.5 Find duplicate ID record
    Check all record that contain duplicate IDs

In [12]:
AimoScore.groupby("ID").filter(lambda x: len(x) > 1)

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
2088   0.527640              0.575323              0.193687   
2089   0.407197              0.279770              0.131038   
2090   0.965930              0.564802              0.311813   
2091   0.527640              0.404489              0.214422   
2092   0.407197              0.606972              0.863897   
2093   0.965930              0.479943              0.250239   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
2088              0.780488              0.956958              0.740794   
2089              0.346724              0.317073              0.945481   
2090              0.376375              0.188905              0.187948   
2091              0.822827              0.937440              0.739255   
2092              0.467526              0.225406              0.946036   
2093              0.366762              0.316141              0.489494   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
2088              0.912482              0.630799              0.632233   
2089              0.646581              0.913439              0.521760   
2090              0.347202              0.227642              0.521760   
2091              0.909265              0.681471              0.658071   
2092              0.706781              0.934097              0.303247   
2093              0.444126              0.681471              0.303247   

      No_9_Angle_Deviation  ...  No_20_NASM_Deviation  No_21_NASM_Deviation  \
2088              0.708274  ...              0.656624              0.642276   
2089              0.570540  ...              0.945002              0.642276   
2090              0.248685  ...              0.753228              0.642276   
2091              0.822350  ...              0.886819              0.640401   
2092              0.576409  ...              0.886819              0.640401   
2093              0.266953  ...              0.886819              0.640401   

      No_22_NASM_Deviation  No_23_NASM_Deviation  No_24_NASM_Deviation  \
2088              0.552846              0.648972              0.578192   
2089              0.596365              0.648972              0.749880   
2090              0.552846              0.648972              0.869440   
2091              0.525310              0.776982              0.822827   
2092              0.725883              0.757402              0.822827   
2093              0.525310              0.757402              0.877746   

      No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
2088              0.736011             0.764706             0.811095   
2089              0.943568             0.684840             0.711621   
2090              0.316117             0.148733             0.151124   
2091              0.367717             0.584050             0.570201   
2092              0.943649             0.953200             0.956543   
2093              0.444604             0.419293             0.423591   

      EstimatedScore                                           ID  
2088        0.911047  ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect  
2089        0.877571  ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect  
2090        0.030129  ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect  
2091        0.613658  ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect  
2092        0.761700  ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect  
2093        0.042025  ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect  

[6 rows x 43 columns]

In [13]:
WeakLinks.groupby("ID").filter(lambda x: len(x) > 1)

Empty DataFrame
Columns: [ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises]
Index: []

### <font color='green'>We found duplicate ID records in AimoScore</font>

# <font color='green'>Result</font>

    AimoScore has 2094 records with 2091 unique IDs, and 3 Repeating IDs
        - ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect at position 2088, 2091
        - ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect at postion 2089, 2092	
        - ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect at postion 2090, 2093
    
    WeakLinks has 2096 records with 2096 unique IDs
    
    AimoScore's all 2091 (remember 3 IDs areduplicate) IDs have matching IDs in WeakLinks
     
    WeakLinks has 5 extra IDs
        - 076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect at position 277
        - ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect at position 809     
        - 1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect at position 953
        - 18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect at position 1149
        - 49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect at position 1201
        
    AimoScore's duplicate IDs data/record (featrues) values are differnet (Only IDs are repeating). and we are not sure that all IDs need to be unique OR we don't have any other crietria to drop one of duplicatie IDs record. so we will keep these and will find WeakLink's as it is.
        
        


# 3 Join DataSet

## 3.1 Join DataSet
    Join DataSet using ID
    Check for any NAN values

In [14]:
AimoScoreWeakLinks = pd.merge(AimoScore, WeakLinks,on=["ID"])
AimoScoreWeakLinks.describe()

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2094.000000           2094.000000           2094.000000   
mean      0.572740              0.539248              0.504462   
std       0.236510              0.239284              0.282136   
min       0.010000              0.279770              0.092300   
25%       0.407907              0.279770              0.250149   
50%       0.617248              0.500000              0.500000   
75%       0.759378              0.749761              0.750239   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.560258              0.518065              0.513932   
std                0.217828              0.264104              0.270076   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250239              0.250717   
50%                0.500000              0.500000              0.500478   
75%                0.750239              0.750239              0.750239   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.510336              0.519434              0.635923   
std                0.274473              0.262884              0.153280   
min                0.140124              0.194165              0.303247   
25%                0.250717              0.250717              0.521760   
50%                0.500478              0.500956              0.521760   
75%                0.750239              0.750239              0.749761   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  RightShoulderElevation  \
count           2094.000000  ...             2094.000000   
mean               0.512548  ...                0.465277   
std                0.271193  ...                0.475690   
min                0.156385  ...                0.000000   
25%                0.250717  ...                0.030000   
50%                0.500478  ...                0.325000   
75%                0.750239  ...                0.857500   
max                1.000000  ...                1.500000   

       ExcessiveForwardLean  LeftAsymmetricalWeightShift  \
count           2094.000000                  2094.000000   
mean               0.201098                     0.175435   
std                0.418949                     0.397713   
min                0.000000                     0.000000   
25%                0.000000                     0.000000   
50%                0.000000                     0.000000   
75%                0.080000                     0.050000   
max                1.500000                     1.500000   

       RightAsymmetricalWeightShift  LeftKneeMovesInward  \
count                   2094.000000          2094.000000   
mean                       0.065377             0.037693   
std                        0.252373             0.140828   
min                        0.000000             0.000000   
25%                        0.000000             0.000000   
50%                        0.000000             0.000000   
75%                        0.000000             0.000000   
max                        1.500000             1.310000   

       RightKneeMovesInward  LeftKneeMovesOutward  RightKneeMovesOutward  \
count           2094.000000           2094.000000            2094.000000   
mean               0.179255              0.170941               0.238424   
std                0.329244              0.401712               0.528395   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000          

In [15]:
AimoScoreWeakLinks[AimoScoreWeakLinks.isna().any(axis=1)]

Empty DataFrame
Columns: [AimoScore, No_1_Angle_Deviation, No_2_Angle_Deviation, No_3_Angle_Deviation, No_4_Angle_Deviation, No_5_Angle_Deviation, No_6_Angle_Deviation, No_7_Angle_Deviation, No_8_Angle_Deviation, No_9_Angle_Deviation, No_10_Angle_Deviation, No_11_Angle_Deviation, No_12_Angle_Deviation, No_13_Angle_Deviation, No_1_NASM_Deviation, No_2_NASM_Deviation, No_3_NASM_Deviation, No_4_NASM_Deviation, No_5_NASM_Deviation, No_6_NASM_Deviation, No_7_NASM_Deviation, No_8_NASM_Deviation, No_9_NASM_Deviation, No_10_NASM_Deviation, No_11_NASM_Deviation, No_12_NASM_Deviation, No_13_NASM_Deviation, No_14_NASM_Deviation, No_15_NASM_Deviation, No_16_NASM_Deviation, No_17_NASM_Deviation, No_18_NASM_Deviation, No_19_NASM_Deviation, No_20_NASM_Deviation, No_21_NASM_Deviation, No_22_NASM_Deviation, No_23_NASM_Deviation, No_24_NASM_Deviation, No_25_NASM_Deviation, No_1_Time_Deviation, No_2_Time_Deviation, EstimatedScore, ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises]
Index: []

[0 rows x 59 columns]

## 3.2 Find Weakest link
    Finding the weak link using weak link columns(14) that has maximum score in each row and adding it as a new column 'WeakLinks'

In [16]:
AimoScoreWeakLinks["WeakLinks"] = AimoScoreWeakLinks.loc[:,"ForwardHead":"RightHeelRises"].idxmax(axis=1)

In [17]:
cols = AimoScoreWeakLinks.columns
quantative_cols = AimoScoreWeakLinks._get_numeric_data().columns
qualitative_cols = list(set(cols) - set(quantative_cols))

print('\nquantative predictors/variables')
for index, col in enumerate(quantative_cols, start=1):
   print (index, '.\'', col, '\' ', sep='', end='')

print('\n\nqualitative predictors/variables')
for index, col in enumerate(qualitative_cols, start=1):
   print (index, '.\'', col, '\' ', sep='', end='')


quantative predictors/variables
1.'AimoScore' 2.'No_1_Angle_Deviation' 3.'No_2_Angle_Deviation' 4.'No_3_Angle_Deviation' 5.'No_4_Angle_Deviation' 6.'No_5_Angle_Deviation' 7.'No_6_Angle_Deviation' 8.'No_7_Angle_Deviation' 9.'No_8_Angle_Deviation' 10.'No_9_Angle_Deviation' 11.'No_10_Angle_Deviation' 12.'No_11_Angle_Deviation' 13.'No_12_Angle_Deviation' 14.'No_13_Angle_Deviation' 15.'No_1_NASM_Deviation' 16.'No_2_NASM_Deviation' 17.'No_3_NASM_Deviation' 18.'No_4_NASM_Deviation' 19.'No_5_NASM_Deviation' 20.'No_6_NASM_Deviation' 21.'No_7_NASM_Deviation' 22.'No_8_NASM_Deviation' 23.'No_9_NASM_Deviation' 24.'No_10_NASM_Deviation' 25.'No_11_NASM_Deviation' 26.'No_12_NASM_Deviation' 27.'No_13_NASM_Deviation' 28.'No_14_NASM_Deviation' 29.'No_15_NASM_Deviation' 30.'No_16_NASM_Deviation' 31.'No_17_NASM_Deviation' 32.'No_18_NASM_Deviation' 33.'No_19_NASM_Deviation' 34.'No_20_NASM_Deviation' 35.'No_21_NASM_Deviation' 36.'No_22_NASM_Deviation' 37.'No_23_NASM_Deviation' 38.'No_24_NASM_Deviation' 39.'

In [18]:
AimoScoreWeakLinks["WeakLinks"].value_counts()

LeftArmFallForward              630
RightArmFallForward             472
RightKneeMovesOutward           277
RightShoulderElevation          250
ForwardHead                     115
ExcessiveForwardLean            100
LeftAsymmetricalWeightShift      73
LeftShoulderElevation            58
LeftKneeMovesOutward             49
RightKneeMovesInward             45
RightAsymmetricalWeightShift     19
LeftKneeMovesInward               3
LeftHeelRises                     2
RightHeelRises                    1
Name: WeakLinks, dtype: int64

## 3.3 Discard extra columns
    Discard the 14 weak link scores from the joint table

In [19]:
AimoScoreWeakLinks.drop(AimoScoreWeakLinks.columns.to_series()["ForwardHead":"RightHeelRises"], axis=1, inplace=True)
len(AimoScoreWeakLinks.columns)

46

In [20]:
AimoScoreWeakLinks.describe()

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2094.000000           2094.000000           2094.000000   
mean      0.572740              0.539248              0.504462   
std       0.236510              0.239284              0.282136   
min       0.010000              0.279770              0.092300   
25%       0.407907              0.279770              0.250149   
50%       0.617248              0.500000              0.500000   
75%       0.759378              0.749761              0.750239   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.560258              0.518065              0.513932   
std                0.217828              0.264104              0.270076   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250239              0.250717   
50%                0.500000              0.500000              0.500478   
75%                0.750239              0.750239              0.750239   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.510336              0.519434              0.635923   
std                0.274473              0.262884              0.153280   
min                0.140124              0.194165              0.303247   
25%                0.250717              0.250717              0.521760   
50%                0.500478              0.500956              0.521760   
75%                0.750239              0.750239              0.749761   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  No_20_NASM_Deviation  No_21_NASM_Deviation  \
count           2094.000000  ...           2094.000000           2094.000000   
mean               0.512548  ...              0.715943              0.706319   
std                0.271193  ...              0.100437              0.105886   
min                0.156385  ...              0.656624              0.640401   
25%                0.250717  ...              0.656624              0.642276   
50%                0.500478  ...              0.656624              0.642276   
75%                0.750239  ...              0.751196              0.749761   
max                1.000000  ...              1.000000              1.000000   

       No_22_NASM_Deviation  No_23_NASM_Deviation  No_24_NASM_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.652969              0.710774              0.667517   
std                0.141000              0.103196              0.130989   
min                0.525310              0.648972              0.578192   
25%                0.552846              0.648972              0.578192   
50%                0.552846              0.648972              0.578192   
75%                0.749880              0.751196              0.751196   
max                1.000000              1.000000              1.000000   

       No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
count           2094.000000          2094.000000          2094.000000   
mean               0.548089             0.512021             0.512373   
std                0.230642             0.273056             0.272647   
min                0.308943             0.148733             0.151124   
25%                0.308943             0.250717             0.250717   
50%                0.500000             0.500478             0.500478   
75%                0.750239             0.750239             0.750239   
max                1.000000             1.000000             1.000000   

       

In [21]:
AimoScoreWeakLinks

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
0      0.323667              0.538020              0.815878   
1      0.323699              0.443807              0.306552   
2      0.848327              0.603539              0.373984   
3      0.351332              0.484935              0.623625   
4      0.627181              0.860832              0.657580   
...         ...                   ...                   ...   
2089   0.527640              0.404489              0.214422   
2090   0.407197              0.279770              0.131038   
2091   0.407197              0.606972              0.863897   
2092   0.965930              0.564802              0.311813   
2093   0.965930              0.479943              0.250239   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
0                 0.346724              0.382114              0.302248   
1                 0.823529              0.188905              0.497370   
2                 0.346724              0.590626              0.341942   
3                 0.380201              0.975132              0.509326   
4                 0.745576              0.552846              0.375897   
...                    ...                   ...                   ...   
2089              0.822827              0.937440              0.739255   
2090              0.346724              0.317073              0.945481   
2091              0.467526              0.225406              0.946036   
2092              0.376375              0.188905              0.187948   
2093              0.366762              0.316141              0.489494   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
0                 0.947872              0.275945              0.521760   
1                 0.140124              0.664275              0.521760   
2                 0.298900              0.276901              0.623625   
3                 0.888570              0.363462              0.847441   
4                 0.483022              0.388331              0.521760   
...                    ...                   ...                   ...   
2089              0.909265              0.681471              0.658071   
2090              0.646581              0.913439              0.521760   
2091              0.706781              0.934097              0.303247   
2092              0.347202              0.227642              0.521760   
2093              0.444126              0.681471              0.303247   

      No_9_Angle_Deviation  ...  No_23_NASM_Deviation  No_24_NASM_Deviation  \
0                 0.457198  ...              0.648972              0.578192   
1                 0.729316  ...              0.848876              0.889048   
2                 0.658058  ...              0.648972              0.578192   
3                 0.237207  ...              0.648972              0.578192   
4                 0.387853  ...              0.648972              0.578192   
...                    ...  ...                   ...                   ...   
2089              0.822350  ...              0.776982              0.822827   
2090              0.570540  ...              0.648972              0.749880   
2091              0.576409  ...              0.757402              0.822827   
2092              0.248685  ...              0.648972              0.869440   
2093              0.266953  ...              0.757402              0.877746   

      No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
0                 0.560019             0.821616             0.818747   
1                 0.816834             0.307987             0.248207   
2                 0.555715             0.218556             0.235294   
3                 0.744620             0.458154             0.432807   
4                 0.308943             0.805356             0.774271   
...                    ...                  ...                  ...   
2089              

## 4 Classification

### <font color='green'>38 Predictors</font>

Use 80% with 38 predictors (all angle deviations and nasm deviations)

In [22]:
trainingPercent = 0.8

y_train, y_test ,X_train, X_test  = train_test_split(AimoScoreWeakLinks[["WeakLinks"]], AimoScoreWeakLinks.drop(columns=["WeakLinks","AimoScore","SCORE","ID","Date","No_1_Time_Deviation","No_2_Time_Deviation","EstimatedScore"]), train_size = trainingPercent)

### Logestic Regression

In [23]:
lr_ovr = LogisticRegression()
lr_ovr.fit(X_train,y_train)
#print(lr_ovr.coef_)
#print(lr_ovr.coef_.shape)
#print(lr_ovr.intercept_)
#print(lr_ovr.intercept_.shape)
lr_ovr_pred = lr_ovr.predict(X_test)
#print(confusion_matrix(y_test,lr_ovr_pred))
print(classification_report(y_test, lr_ovr_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_ovr_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.741     0.606     0.667        33
                 ForwardHead      0.476     0.556     0.513        36
          LeftArmFallForward      0.527     0.756     0.621       180
 LeftAsymmetricalWeightShift      0.889     0.286     0.432        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.429     0.188     0.261        16
       LeftShoulderElevation      0.750     0.176     0.286        17
         RightArmFallForward      0.518     0.431     0.470       137
RightAsymmetricalWeightShift      0.000     0.000     0.000         4
        RightKneeMovesInward      0.000     0.000     0.000        10
       RightKneeMovesOutward      0.892     0.871     0.881        85
      RightShoulderElevation      0.518     0.543     0.530        81

                  

In [24]:
lr_mn = LogisticRegression()
lr_mn.fit(X_train,y_train)
#print(lr_mn.coef_)
#print(lr_mn.coef_.shape)
#print(lr_mn.intercept_)
#print(lr_mn.intercept_.shape)
lr_mn_pred = lr_mn.predict(X_test)
#print(confusion_matrix(y_test,lr_mn_pred))
print(classification_report(y_test, lr_mn_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_mn_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.741     0.606     0.667        33
                 ForwardHead      0.476     0.556     0.513        36
          LeftArmFallForward      0.527     0.756     0.621       180
 LeftAsymmetricalWeightShift      0.889     0.286     0.432        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.429     0.188     0.261        16
       LeftShoulderElevation      0.750     0.176     0.286        17
         RightArmFallForward      0.518     0.431     0.470       137
RightAsymmetricalWeightShift      0.000     0.000     0.000         4
        RightKneeMovesInward      0.000     0.000     0.000        10
       RightKneeMovesOutward      0.892     0.871     0.881        85
      RightShoulderElevation      0.518     0.543     0.530        81

                  

### Naive Bayes Classification

In [25]:
gnb = GaussianNB()

gnb_pred = gnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, gnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, gnb_pred)))

Accuracy:0.4562798092209857


In [26]:
mnb = MultinomialNB()

mnb_pred = mnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, mnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, mnb_pred)))

Accuracy:0.3942766295707472


### LDA

In [27]:
lda = LinearDiscriminantAnalysis()
lda_model = lda.fit(X_train, y_train)
#print(lda_model.priors_)
#print(lda_model.means_)
#print(lda_model.coef_)
lda_model_pred = lda_model.predict(X_test)
#print(confusion_matrix(y_test, lda_model_pred))
print(classification_report(y_test, lda_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lda_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.595     0.667     0.629        33
                 ForwardHead      0.478     0.611     0.537        36
          LeftArmFallForward      0.575     0.678     0.622       180
 LeftAsymmetricalWeightShift      0.458     0.393     0.423        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.267     0.250     0.258        16
       LeftShoulderElevation      0.182     0.118     0.143        17
         RightArmFallForward      0.529     0.460     0.492       137
RightAsymmetricalWeightShift      0.000     0.000     0.000         4
        RightKneeMovesInward      0.143     0.100     0.118        10
       RightKneeMovesOutward      0.962     0.882     0.920        85
      RightShoulderElevation      0.438     0.395     0.416        81

                  

### K-Nearest Neighbors Clustering K = 1 2 3

In [28]:
knn = neighbors.KNeighborsClassifier(n_neighbors=1)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.344     0.333     0.338        33
                 ForwardHead      0.455     0.417     0.435        36
          LeftArmFallForward      0.509     0.467     0.487       180
 LeftAsymmetricalWeightShift      0.462     0.429     0.444        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.200     0.188     0.194        16
       LeftShoulderElevation      0.071     0.059     0.065        17
         RightArmFallForward      0.415     0.496     0.452       137
RightAsymmetricalWeightShift      0.167     0.250     0.200         4
        RightKneeMovesInward      0.250     0.300     0.273        10
       RightKneeMovesOutward      0.875     0.906     0.890        85
      RightShoulderElevation      0.554     0.506     0.529        81

                  

In [29]:
knn = neighbors.KNeighborsClassifier(n_neighbors=2)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.320     0.485     0.386        33
                 ForwardHead      0.333     0.472     0.391        36
          LeftArmFallForward      0.479     0.622     0.541       180
 LeftAsymmetricalWeightShift      0.348     0.286     0.314        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.364     0.250     0.296        16
       LeftShoulderElevation      0.136     0.176     0.154        17
         RightArmFallForward      0.445     0.358     0.397       137
RightAsymmetricalWeightShift      0.000     0.000     0.000         4
        RightKneeMovesInward      0.000     0.000     0.000        10
       RightKneeMovesOutward      0.894     0.894     0.894        85
      RightShoulderElevation      0.771     0.333     0.466        81

                  

In [30]:
knn = neighbors.KNeighborsClassifier(n_neighbors=3)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.356     0.485     0.410        33
                 ForwardHead      0.351     0.556     0.430        36
          LeftArmFallForward      0.457     0.533     0.492       180
 LeftAsymmetricalWeightShift      0.381     0.286     0.327        28
               LeftHeelRises      0.000     0.000     0.000         1
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.429     0.188     0.261        16
       LeftShoulderElevation      0.091     0.059     0.071        17
         RightArmFallForward      0.433     0.423     0.428       137
RightAsymmetricalWeightShift      0.000     0.000     0.000         4
        RightKneeMovesInward      0.333     0.100     0.154        10
       RightKneeMovesOutward      0.915     0.882     0.898        85
      RightShoulderElevation      0.593     0.432     0.500        81

                  

### <font color='green'>26 Predictors</font>

Use 80% with 26 predictors (all angle deviations and nasm deviations - symmetry)

In [31]:
trainingPercent = 0.8

y_train, y_test ,X_train, X_test  = train_test_split(AimoScoreWeakLinks[["WeakLinks"]], AimoScoreWeakLinks.drop(columns=["WeakLinks","AimoScore","SCORE","ID","Date","No_6_Angle_Deviation","No_7_Angle_Deviation","No_11_Angle_Deviation","No_12_Angle_Deviation","No_13_Angle_Deviation","No_2_NASM_Deviation","No_5_NASM_Deviation","No_9_NASM_Deviation","No_12_NASM_Deviation","No_14_NASM_Deviation","No_16_NASM_Deviation","No_19_NASM_Deviation","No_2_NASM_Deviation","No_1_Time_Deviation","No_2_Time_Deviation","EstimatedScore"]), train_size = trainingPercent)

### Logestic Regression

In [32]:
lr_ovr = LogisticRegression()
lr_ovr.fit(X_train,y_train)
#print(lr_ovr.coef_)
#print(lr_ovr.coef_.shape)
#print(lr_ovr.intercept_)
#print(lr_ovr.intercept_.shape)
lr_ovr_pred = lr_ovr.predict(X_test)
#print(confusion_matrix(y_test,lr_ovr_pred))
print(classification_report(y_test, lr_ovr_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_ovr_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.562     0.600     0.581        15
                 ForwardHead      0.500     0.286     0.364         7
          LeftArmFallForward      0.600     0.600     0.600        70
 LeftAsymmetricalWeightShift      0.500     0.250     0.333        12
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.000     0.000     0.000         3
       LeftShoulderElevation      0.000     0.000     0.000         4
         RightArmFallForward      0.360     0.450     0.400        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.794     0.871     0.831        31
      RightShoulderElevation      0.407     0.550     0.468        20

                    accuracy                          0.533       210
                  

In [33]:
lr_mn = LogisticRegression()
lr_mn.fit(X_train,y_train)
#print(lr_mn.coef_)
#print(lr_mn.coef_.shape)
#print(lr_mn.intercept_)
#print(lr_mn.intercept_.shape)
lr_mn_pred = lr_mn.predict(X_test)
#print(confusion_matrix(y_test,lr_mn_pred))
print(classification_report(y_test, lr_mn_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_mn_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.562     0.600     0.581        15
                 ForwardHead      0.500     0.286     0.364         7
          LeftArmFallForward      0.600     0.600     0.600        70
 LeftAsymmetricalWeightShift      0.500     0.250     0.333        12
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.000     0.000     0.000         3
       LeftShoulderElevation      0.000     0.000     0.000         4
         RightArmFallForward      0.360     0.450     0.400        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.794     0.871     0.831        31
      RightShoulderElevation      0.407     0.550     0.468        20

                    accuracy                          0.533       210
                  

### Naive Bayes Classification

In [34]:
gnb = GaussianNB()

gnb_pred = gnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, gnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, gnb_pred)))

Accuracy:0.48095238095238096


In [35]:
mnb = MultinomialNB()

mnb_pred = mnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, mnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, mnb_pred)))

Accuracy:0.4238095238095238


### LDA

In [36]:
lda = LinearDiscriminantAnalysis()
lda_model = lda.fit(X_train, y_train)
#print(lda_model.priors_)
#print(lda_model.means_)
#print(lda_model.coef_)
lda_model_pred = lda_model.predict(X_test)
#print(confusion_matrix(y_test, lda_model_pred))
print(classification_report(y_test, lda_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lda_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.667     0.667     0.667        15
                 ForwardHead      0.500     0.571     0.533         7
          LeftArmFallForward      0.651     0.586     0.617        70
 LeftAsymmetricalWeightShift      0.750     0.500     0.600        12
               LeftHeelRises      0.000     0.000     0.000         0
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.200     0.333     0.250         3
       LeftShoulderElevation      0.000     0.000     0.000         4
         RightArmFallForward      0.455     0.500     0.476        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.875     0.903     0.889        31
      RightShoulderElevation      0.407     0.550     0.468        20

                  

### K-Nearest Neighbors Clustering K = 1 2 3

In [37]:
knn = neighbors.KNeighborsClassifier(n_neighbors=1)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.800     0.533     0.640        15
                 ForwardHead      0.400     0.571     0.471         7
          LeftArmFallForward      0.561     0.457     0.504        70
 LeftAsymmetricalWeightShift      0.667     0.167     0.267        12
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.250     0.333     0.286         3
       LeftShoulderElevation      0.200     0.250     0.222         4
         RightArmFallForward      0.345     0.475     0.400        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.933     0.903     0.918        31
      RightShoulderElevation      0.303     0.500     0.377        20

                    accuracy                          0.500       210
                  

In [38]:
knn = neighbors.KNeighborsClassifier(n_neighbors=2)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.550     0.733     0.629        15
                 ForwardHead      0.278     0.714     0.400         7
          LeftArmFallForward      0.519     0.586     0.550        70
 LeftAsymmetricalWeightShift      0.571     0.333     0.421        12
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.333     0.333     0.333         3
       LeftShoulderElevation      0.200     0.250     0.222         4
         RightArmFallForward      0.412     0.350     0.378        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.966     0.903     0.933        31
      RightShoulderElevation      0.182     0.100     0.129        20

                    accuracy                          0.510       210
                  

In [39]:
knn = neighbors.KNeighborsClassifier(n_neighbors=3)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.632     0.800     0.706        15
                 ForwardHead      0.143     0.286     0.190         7
          LeftArmFallForward      0.547     0.586     0.566        70
 LeftAsymmetricalWeightShift      0.500     0.333     0.400        12
               LeftHeelRises      0.000     0.000     0.000         0
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.500     0.333     0.400         3
       LeftShoulderElevation      0.200     0.250     0.222         4
         RightArmFallForward      0.429     0.375     0.400        40
RightAsymmetricalWeightShift      0.000     0.000     0.000         2
        RightKneeMovesInward      0.000     0.000     0.000         5
       RightKneeMovesOutward      0.933     0.903     0.918        31
      RightShoulderElevation      0.211     0.200     0.205        20

                  

### QDA

In [40]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

qda = QuadraticDiscriminantAnalysis()
qda_model = qda.fit(X_train, y_train)
print(qda_model.priors_)
print(qda_model.means_)
#print(qda_model.coef_)
#qda_model_pred = qda_model.predict(X_test)
#print(confusion_matrix(y_test, qda_model_pred))
#print(classification_report(y_test, qda_model_pred, digits = 3))
#print('Accuracy:' + str(accuracy_score(y_test, qda_model_pred)))

ValueError: y has only 1 sample in class RightHeelRises, covariance is ill defined.